File: framework/ui/src/io/intino/alexandria/ui/displays/components/Number.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.notifiers.NumberNotifier;

public class Number<DN extends NumberNotifier, B extends Box> extends AbstractNumber<DN, B> {
    private Double value;
    private Double min;
    private Double max;
    private Double step;

    public Number(B box) {
        super(box);
    }

    public Double value() {
        return value;
    }

    public Double min() {
        return min;
    }

    public Double max() {
        return max;
    }

    public Double step() {
        return step;
    }

    public void value(double value) {
        value(Double.valueOf(value));
    }

    public void value(Double value) {
        _value(value);
        notifier.refresh(value);
    }

    protected Number _value(double value) {
        return _value(Double.valueOf(value));
    }

    protected Number _value(Double value) {
        this.value = value;
        return this;
    }

    protected Number _min(double min) {
        return _min(Double.valueOf(min));
    }

    protected Number _min(Double min) {
        this.min = min;
        return this;
    }

    protected Number _max(double min) {
        return _max(Double.valueOf(min));
    }

    protected Number _max(Double max) {
        this.max = max;
        return this;
    }

    protected Number _step(double step) {
        return _step(Double.valueOf(step));
    }

    protected Number _step(Double step) {
        this.step = step;
        return this;
    }

}

File: framework/ui-elements/src/displays/components/Number.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import Typography from "@material-ui/core/Typography";
import Popover from "@material-ui/core/Popover";
import AbstractNumber from "../../../gen/displays/components/AbstractNumber";
import NumberNotifier from "../../../gen/displays/notifiers/NumberNotifier";
import NumberRequester from "../../../gen/displays/requesters/NumberRequester";
import TextBehavior from "./behaviors/TextBehavior";
import Block from "./Block";
import DisplayFactory from "alexandria-ui-elements/src/displays/DisplayFactory";
import NumberUtil from "alexandria-ui-elements/src/util/NumberUtil";
import ComponentBehavior from "./behaviors/ComponentBehavior";
import classnames from "classnames";

const styles = theme => ({
	label: {
		color: theme.palette.grey.A700,
		marginRight: "5px"
	},
	value : {
		margin: "0 !important"
	},
	prefix : {
		color: theme.palette.grey.A700,
		fontSize: "10pt",
//		marginTop: "2px",
		marginRight: "3px"
	},
	suffix : {
		color: theme.palette.grey.A700,
		fontSize: "10pt",
		marginLeft: "3px"
	},
	withExpanded : {
		cursor: "pointer",
		color: theme.palette.primary.main,
	}
});

class Number extends AbstractNumber {

	constructor(props) {
		super(props);
		this.notifier = new NumberNotifier(this);
		this.requester = new NumberRequester(this);
		this.state = {
			...this.state,
		}
	};

	render() {
		const { classes } = this.props;
		const value = this.state.value;
		const variant = this.variant("body1");
		const expanded = this.state.expanded;
		const format = expanded ? this._defaultFormat() : this._format();
		const expandedClass = this.props.style != null ? classes.withExpanded : undefined;

		if (value == null || value === "" || !this.state.visible) return (<React.Fragment/>);

		return (
			<Block layout="horizontal center" style={this.style()}>
				{ ComponentBehavior.labelBlock(this.props, 'body1', {...this.style(),margin:'0 5px 0 0'}) }
				{this.state.prefix !== undefined ? <Typography variant={variant} className={classes.prefix}>{this.state.prefix}</Typography> : undefined }
				<Typography className={classnames(expandedClass, classes.value)} onClick={this.handleToggleExpanded.bind(this)} variant={variant} style={this.style()}>{NumberUtil.format(value, this.translate(format))}</Typography>
				{ this.state.suffix !== undefined ? <Typography variant={variant} className={classes.suffix}>{this.state.suffix}</Typography> : undefined }
			</Block>
		);
	};

	refresh = (value) => {
		this.setState({ value });
	};

	_format = () => {
		const style = this.props.style;
		if (style == null) return this._defaultFormat();
		const format = this._addDecimals("0");
		if (style === "Currency") return format + "a";
		else if (style === "Bytes") return format + "b";
		else if (style === "Percentage" && this.state.suffix == null) return "(" + format + "%)";
		else if (style === "Exponential" && this.state.suffix == null) return "(" + format + "e+0)";
		return this._defaultFormat();
	};

	_defaultFormat = () => {
		return this._addDecimals("0,0");
	};

	_addDecimals = (format) => {
		const decimals = this.state.decimals != null ? this.state.decimals : 0;
		if (decimals > 0) format += ".";
		for (let i=0; i<decimals; i++) format += "0";
		return format;
	};

	handleToggleExpanded = () => {
		this.setState({ expanded: !this.state.expanded });
	};

}

export default withStyles(styles, { withTheme: true })(Number);
DisplayFactory.register("Number", withStyles(styles, { withTheme: true })(Number));